File: ruby/lib/cucumber/tag_expressions/parser.rb
# frozen_string_literal: true

require 'cucumber/tag_expressions/expressions'

module Cucumber
  module TagExpressions
    class Parser
      def initialize
        @expressions = []
        @operators = []
      end

      def parse(infix_expression)
        expected_token_type = :operand
        tokens = tokenize(infix_expression)
        return True.new if tokens.empty?

        tokens.each { |token| expected_token_type = handle_sequential_tokens(token, infix_expression, expected_token_type) }
        while @operators.any?
          raise "Tag expression \"#{infix_expression}\" could not be parsed because of syntax error: Unmatched (." if @operators.last == '('
          push_expression(infix_expression, @operators.pop)
        end
        @expressions.pop
      end

      private

      def lower_precedence?(operation)
        (operator_types.dig(operation, :assoc) == :left && precedence(operation) <= precedence(@operators.last)) ||
          (operator_types.dig(operation, :assoc) == :right && precedence(operation) < precedence(@operators.last))
      end

      def operator?(token)
        %i[unary_operator binary_operator].include?(operator_types.dig(token, :type))
      end

      def precedence(token)
        operator_types.dig(token, :precedence)
      end

      def tokenize(infix_expression)
        tokens = []
        escaped = false
        token = +''
        infix_expression.chars.each do |char|
          if escaped
            unless char == '(' || char == ')' || char == '\\' || whitespace?(char)
              raise "Tag expression \"#{infix_expression}\" could not be parsed because of syntax error: Illegal escape before \"#{char}\"."
            end

            token += char
            escaped = false
          elsif char == '\\'
            escaped = true
          elsif char == '(' || char == ')' || whitespace?(char)
            unless token.empty?
              tokens.push(token)
              token = +''
            end
            tokens.push(char) unless whitespace?(char)
          else
            token += char
          end
        end
        tokens.push(token) unless token.empty?
        tokens
      end

      def push_expression(infix_expression, token)
        case token
        when 'and' then @expressions.push(And.new(*pop(infix_expression, @expressions, 2)))
        when 'or'  then @expressions.push(Or.new(*pop(infix_expression, @expressions, 2)))
        when 'not' then @expressions.push(Not.new(pop(infix_expression, @expressions)))
        else            @expressions.push(Literal.new(token))
        end
      end

      def handle_sequential_tokens(token, infix_expression, expected_token_type)
        if operator_types[token]
          send("handle_#{operator_types.dig(token, :type)}", infix_expression, token, expected_token_type)
        else
          handle_literal(infix_expression, token, expected_token_type)
        end
      end

      def handle_unary_operator(infix_expression, token, expected_token_type)
        check(infix_expression, expected_token_type, :operand)
        @operators.push(token)
        :operand
      end

      def handle_binary_operator(infix_expression, token, expected_token_type)
        check(infix_expression, expected_token_type, :operator)
        push_expression(infix_expression, @operators.pop) while @operators.any? && operator?(@operators.last) && lower_precedence?(token)
        @operators.push(token)
        :operand
      end

      def handle_open_paren(infix_expression, token, expected_token_type)
        check(infix_expression, expected_token_type, :operand)
        @operators.push(token)
        :operand
      end

      def handle_close_paren(infix_expression, _token, expected_token_type)
        check(infix_expression, expected_token_type, :operator)
        push_expression(infix_expression, @operators.pop) while @operators.any? && @operators.last != '('
        raise "Tag expression \"#{infix_expression}\" could not be parsed because of syntax error: Unmatched )." if @operators.empty?

        @operators.pop if @operators.last == '('
        :operator
      end

      def handle_literal(infix_expression, token, expected_token_type)
        check(infix_expression, expected_token_type, :operand)
        push_expression(infix_expression, token)
        :operator
      end

      def check(infix_expression, expected_token_type, token_type)
        return if expected_token_type == token_type

        raise "Tag expression \"#{infix_expression}\" could not be parsed because of syntax error: Expected #{expected_token_type}."
      end

      def pop(infix_expression, array, amount = 1)
        result = array.pop(amount)
        raise "Tag expression \"#{infix_expression}\" could not be parsed because of syntax error: Expression is incomplete." if result.length != amount

        amount == 1 ? result.first : result
      end

      def operator_types
        {
          'or' => { type: :binary_operator, precedence: 0, assoc: :left },
          'and' => { type: :binary_operator, precedence: 1, assoc: :left },
          'not' => { type: :unary_operator, precedence: 2, assoc: :right },
          ')' => { type: :close_paren, precedence: -1 },
          '(' => { type: :open_paren, precedence: 1 }
        }
      end

      def whitespace?(char)
        char.match(/\s/)
      end
    end
  end
end

File: cpp/src/cucumber/tag-expressions/Parser.cpp
#include "cucumber/tag-expressions/Parser.hpp"
#include "cucumber/tag-expressions/Error.hpp"
#include "cucumber/tag-expressions/Model.hpp"
#include "cucumber/tag-expressions/Token.hpp"
#include <cctype>
#include <cstddef>
#include <deque>
#include <locale>
#include <memory>
#include <stack>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace cucumber::tag_expressions
{
    namespace
    {
        Error SyntaxError(std::string_view expression, std::string_view reason)
        {
            return Error("Tag expression \"" + std::string{ expression } + "\" could not be parsed because of syntax error: " + std::string{ reason } + ".");
        }

        void EnsureExpectedTokenType(std::string_view expression, TokenType tokenType, TokenType expected)
        {
            if (tokenType != expected)
            {
                throw SyntaxError(expression, "Expected " + std::string{ TokenTypeMap().at(expected) });
            }
        }

        void RequireArgCount(std::string_view expression, std::deque<std::unique_ptr<Expression>>& expressions, std::size_t number)
        {
            if (expressions.size() < number)
            {
                throw SyntaxError(expression, "Expected operand");
            }
        }

        template<class T>
        void PushBinary(std::string_view expression, std::deque<std::unique_ptr<Expression>>& expressions)
        {
            RequireArgCount(expression, expressions, 2);

            auto term2 = std::move(expressions.back());
            expressions.pop_back();
            auto term1 = std::move(expressions.back());
            expressions.pop_back();

            expressions.push_back(std::make_unique<T>(std::move(term1), std::move(term2)));
        }

        template<class T>
        void PushUnary(std::string_view expression, std::deque<std::unique_ptr<Expression>>& expressions)
        {
            RequireArgCount(expression, expressions, 1);

            auto term = std::move(expressions.back());
            expressions.pop_back();

            expressions.push_back(std::make_unique<T>(std::move(term)));
        }

        void PushExpression(std::string_view expression, const Token& token, std::deque<std::unique_ptr<Expression>>& expressions)
        {
            if (token == OR)
            {
                PushBinary<OrExpression>(expression, expressions);
            }
            else if (token == AND)
            {
                PushBinary<AndExpression>(expression, expressions);
            }
            else if (token == NOT)
            {
                PushUnary<NotExpression>(expression, expressions);
            }
            else
            {
                throw Error("(Unexpected token: " + std::string{ token.keyword } + ")");
            }
        }

        std::vector<std::string> Tokenize(std::string_view expression)
        {
            std::vector<std::string> tokens{};
            auto escaped = false;
            std::string token = {};

            for (char ch : expression) // NOLINT(readability-identifier-length)
            {
                if (escaped)
                {
                    if ((ch != '(' && ch != ')' && ch != '\\') && !std::isspace(ch, std::locale()))
                    {
                        throw SyntaxError(expression, "Illegal escape before \"" + std::string{ ch } + "\"");
                    }

                    token += ch;
                    escaped = false;
                }
                else if (ch == '\\')
                {
                    escaped = true;
                }
                else if (ch == '(' || ch == ')' || std::isspace(ch, std::locale()))
                {
                    if (!token.empty())
                    {
                        tokens.push_back(std::move(token));
                        token.clear();
                    }
                    if (ch != ' ')
                    {
                        tokens.push_back(std::string{ ch });
                    }
                }
                else
                {
                    token += ch;
                }
            }

            if (!token.empty())
            {
                tokens.push_back(std::move(token));
            }

            return tokens;
        }

        const Token* SelectToken(std::string_view expression)
        {
            const auto& map = TokenMap();

            if (map.find(expression) == map.end())
            {
                return nullptr;
            }

            return &map.at(expression);
        }
    }

    std::unique_ptr<Expression> Parse(std::string_view expression)
    {
        auto tokens = Tokenize(expression);

        if (tokens.empty())
        {
            return std::make_unique<TrueExpression>();
        }

        std::stack<Token> operations;
        std::deque<std::unique_ptr<Expression>> expressions;
        auto expectedTokenType = TokenType::operand;

        for (auto index = 0; index < tokens.size(); ++index)
        {
            const auto& part = tokens[index];

            if (const auto* token = SelectToken(part); token == nullptr)
            {
                EnsureExpectedTokenType(expression, TokenType::operand, expectedTokenType);
                expressions.push_back(std::make_unique<LiteralExpression>(part));
                expectedTokenType = TokenType::operator_;
            }
            else if (*token == NOT || *token == OPEN_PARENTHESIS)
            {
                EnsureExpectedTokenType(expression, TokenType::operand, expectedTokenType);
                operations.push(*token);
                expectedTokenType = TokenType::operand;
            }
            else if (token->IsOperation())
            {
                EnsureExpectedTokenType(expression, TokenType::operator_, expectedTokenType);
                while (!operations.empty() && operations.top().IsOperation() && token->HasLowerPrecedenceThan(operations.top()))
                {
                    auto lastOperation = operations.top();
                    operations.pop();
                    PushExpression(expression, lastOperation, expressions);
                }
                operations.push(*token);
                expectedTokenType = TokenType::operand;
            }
            else if (*token == CLOSE_PARENTHESIS)
            {
                EnsureExpectedTokenType(expression, TokenType::operator_, expectedTokenType);
                while (!operations.empty() && operations.top() != OPEN_PARENTHESIS)
                {
                    auto lastOperation = operations.top();
                    operations.pop();
                    PushExpression(expression, lastOperation, expressions);
                }

                if (operations.empty())
                {
                    throw SyntaxError(expression, "Unmatched )");
                }

                if (operations.top() == OPEN_PARENTHESIS)
                {
                    operations.pop();
                    expectedTokenType = TokenType::operator_;
                }
            }
        }

        while (!operations.empty())
        {
            auto lastOperation = operations.top();
            operations.pop();

            if (lastOperation == OPEN_PARENTHESIS)
            {
                throw SyntaxError(expression, "Unmatched (");
            }

            PushExpression(expression, lastOperation, expressions);
        }

        return std::move(expressions.back());
    }
}
